webhook_service/main.go

```diff
@@ -0,0 +1,873 @@
+		error:   log.New(file, "["+tag+"] [ERROR] ", baseFlags),
+		level:   logLevel,
+	}
+
+	maxQueueSize, err = strconv.ParseInt(os.Getenv("HLL_WH_MAX_QUEUE_LENGTH"), 10, 64)
+	if err != nil {
+		maxQueueSize = 150
+	}
+
+	var maxReattempts int
+	maxReattempts, err = strconv.Atoi(os.Getenv("HLL_WH_MAX_RETRIES"))
+	if err != nil {
+		maxMsgReattempts = 5
+	} else {
+		maxMsgReattempts = maxReattempts
+	}
+
+	var maxSends int
+	maxSends, err = strconv.Atoi(os.Getenv("HLL_LOCAL_MAX_SENDS_PER_SEC"))
+	if err != nil {
+		localRateLimit = 45
+	} else {
+		localRateLimit = maxSends
+	}
+
+	var windowSize int
+	windowSize, err = strconv.Atoi(os.Getenv("HLL_WH_SERVICE_RL_TIME_WINDOW"))
+	if err != nil {
+		rateLimitWindowSize = time.Duration(600 * int(time.Second))
+	} else {
+		rateLimitWindowSize = time.Duration(windowSize * int(time.Second))
+	}
+
+	webhookIDPattern = regexp.MustCompile(`webhooks/([0-9]+)/`)
+	workerLookup = make(map[string]*BucketWorker)
+}
+
+const (
+	inputQueue                 = "discord_webhook_queue:input"
+	firstTimeQueue             = "discord_webhook_queue:first_time"
+	bucketQueuePrefix          = "discord_webhook_queue:bucket:"
+	webhookErrors              = "discord_webhook_queue:webhook_id_errors"
+	transientChannel           = "discord_webhook_transient:channel"
+	message_404                = "discord_webhook:message_404"
+	bucketsHash                = "discord_webhook:webhook_buckets"
+	bucketRateLimitCountHash   = "discord_webhook_bucket_rl_count"
+	discordGloballyRateLimited = "discord_webhook:global_rate_limited"
+)
+
+// WebhookType represents the type of webhook service
+type WebhookType string
+
+const (
+	WebhookTypeDiscord WebhookType = "discord"
+)
+
+// WebhookMessageType represents the underlying type of a webhook message
+type WebhookMessageType string
+
+const (
+	WebhookMessageTypeLogLine     WebhookMessageType = "log_line"
+	WebhookMessageTypeLogLineChat WebhookMessageType = "log_line_chat"
+	WebhookMessageTypeLogLineKill WebhookMessageType = "log_line_kill"
+	WebhookMessageTypeTeamkill    WebhookMessageType = "log_line_teamkill"
+	WebhookMessageTypeAdminPing   WebhookMessageType = "admin_ping"
+	WebhookMessageTypeScoreboard  WebhookMessageType = "scoreboard"
+	WebhookMessageTypeAudit       WebhookMessageType = "audit"
+	WebhookMessageTypeOther       WebhookMessageType = "other"
+)
+
+// DiscordWebhookDict represents the payload for a Discord webhook
+type DiscordWebhookDict struct {
+	URL             string               `json:"url"`
+	WebhookID       string               `json:"webhook_id"`
+	MessageID       *string              `json:"message_id,omitempty"`
+	AllowedMentions *AllowedMentionsDict `json:"allowed_mentions,omitempty"`
+	Content         *string              `json:"content,omitempty"`
+	Embeds          []DiscordEmbedDict   `json:"embeds,omitempty"`
+}
+
+// Override unmarshaling so we can ignore extra fields
+func (d *DiscordWebhookDict) UnmarshalJSON(data []byte) error {
```

Comment:
tbh: I don't get this :D Why do you need to do this? 🤔

Reply:
> tbh: I don't get this :D Why do you need to do this? 🤔

I guess we don't really have to but there's a lot of extra shit that gets serialized into it and we only care about a few very specific things; I didn't want to include all the extra stuff so I just ignored those fields.